### ISSUE TITLE
Intermittent build error with TypeScript v1.5 · TS2388: Function overload must not be static

### ISSUE THREAD

**Issue by mattbrooks2010** [Microsoft/TypeScript#4185]
I am able to reproduce an intermittent build error for multiple Visual Studio projects when using TypeScript v1.5. This was not happening previously with TypeScript v1.4.

```
> tsc -v
> message TS6029: Version 1.5.3
```

The scenario below only occurs during a clean build and only for approximately 1 in 3 builds. This was originally encountered during my CI builds, which invoke the TypeScript compiler via MSBuild. However, I'm also able to reproduce this behaviour when invoking `tsc` with default options via the command line. Again, I am only able to reproduce the behaviour for say 1 in 3 attempts.

The issue seems to be caused by one code file in particular. The rest of the errors appear to be symptoms of that original cause, although the order of the errors would suggest otherwise. All JavaScript output files appear to be properly generated on disk.

Below is just a sample of the symptomatic errors, because there are many of the form:

```
> Index.ts(31,18): error TS2339: Property 'load' does not exist on type 'typeof Ajax'.
> Notifications.ts(4,27): error TS2345: Argument of type 'string' is not assignable to parameter of type 'IAjaxSettings'.
...
```

Here, I believe, is the root cause:

```
> Ajax.ts(57,16): error TS2388: Function overload must not be static.
```

I do not understand why the previous error would only sometimes be raised by the compiler. The overloaded methods we have in `Ajax.ts` have been unchanged for approximately 2 years, since early versions of the TypeScript compiler.

Below is a simplified extract from `Ajax.ts`:

```TypeScript
module App {
    export interface IAjaxSettings {
        url: string;
        data?: any;
    }

    export class Ajax {
        static load(url: string): JQueryXHR;
        static load(options: IAjaxSettings): JQueryXHR;
        static load(options: any) {
            if (typeof options === "string") {
                return Ajax.send("GET", { url: options });
            }

            return Ajax.send("GET", options);
        }

        // Remaining implementation removed
        // ...
    }
}
```

**Comment by danquirk** [comment]
Sounds strange. I'm not sure how much we can do without a larger repro. The code snippet there doesn't have any errors in isolation (modulo defining JQueryHXR and Ajax.send).

**Comment by mhegazy** [comment]
one thing to look for is if any of the input files autogenerated from a previous build step. 

it might be helpful to compare the msbuild log file for failed and succeeded attempts, and make sure that all invocations to tsc.exe are identical, just to rule out other tools in the chain.

**Comment by mattbrooks2010** [comment]
@danquirk Unfortunately no, I cannot share the project. I did spend limited time on producing a minimal repro but couldn't reproduce the intermittent issue in the time available. I will try again.

@mhegazy I don't think this is relevant as I have been able to trigger the issue using tsc.exe directly.

To me though, what is most curious is why I am getting the following error at all:

```
> Ajax.ts(57,16): error TS2388: Function overload must not be static.
```

Are you able to shed any light on what conditions would / should trigger this compiler error?

**Comment by mattbrooks2010** [comment]
I have been able to produce a reasonably minimal repro for this issue. I'd like to share the code privately if possible — what is the best way of doing this?

The repro consists of the smallest number of TypeScript files possible to still reliably produce compiler errors. It contains TypeScript files and a `tsconfig.json` file. There is also a PowerShell script that continuously cleans / builds until it encounters compilation errors (non-zero exit code from `tsc.exe`). On my machine and also a colleague's this usually happens within approximately 20 compilations — most compilation attempts are fine.

Some other information that might be relevant:

OS: Windows 8.1 64 bit
tsc.exe: Version 1.5.3

**Comment by mhegazy** [comment]
feel free to email me. mhegazy at microsoft

**Comment by mattbrooks2010** [comment]
@mhegazy emailed — thanks

**Comment by mhegazy** [comment]
We have tracked this issue to Chakra (tschost.dll). I am working with the Chakra team to see how ans when we can fix this. i will update the issue once i have more information..

for now, node does not seem to show a similar issue. so you can replace references to `tsc.exe` with `node tsc.js`

**Comment by mhegazy** [comment]
We could not get the fix in time for 1.6. Hopefully i have an update with a time line soon.

**Comment by mhegazy** [comment]
TS 1.8 should have a new version of chakra that address the issue. sorry for the delay.
